File: src/mysqlcookbook/recipes/transfer/check_set_value.pl
#!/usr/bin/perl
# check_set_value.pl: Wrapper to demonstrate check_set_value()
# utility function that determines whether a value is a legal
# value of a given SET column.

# Usage: check_set_value.pl db_name tbl_name col_name test_value

use strict;
use warnings;
use Cookbook;
use Cookbook_Utils;

@ARGV == 4 or die "Usage: $0 db_name tbl_name col_name test_value\n";

my $db_name = shift (@ARGV);
my $tbl_name = shift (@ARGV);
my $col_name = shift (@ARGV);
my $val = shift (@ARGV);

my $dbh = Cookbook::connect ();

my $valid = check_set_value ($dbh, $db_name, $tbl_name, $col_name, $val);

print "'$val' "
      . ($valid ? "is" : "is not")
      . " a valid $tbl_name.$col_name value\n";

# For bulk testing of lots of values, get the column metadata,
# construct a hash from the array of SET values, and use that for
# testing values.

#@ _BULK_SET_CHECK_SETUP_
my $ref = get_enumorset_info ($dbh, $db_name, $tbl_name, $col_name);
my %members;
# convert hash key to consistent lettercase
map { $members{lc ($_)} = 1; } @{$ref->{values}};
#@ _BULK_SET_CHECK_SETUP_

# If any element of the test value isn't in the SET,
# the entire value is invalid

#@ _BULK_SET_CHECK_TEST_
$valid = 1;     # assume valid until we find out otherwise
foreach my $elt (split (/,/, lc ($val)))
{
  if (!exists ($members{$elt}))
  {
    $valid = 0; # value contains an invalid element
    last;
  }
}
#@ _BULK_SET_CHECK_TEST_
print "'$val' "
      . ($valid ? "is" : "is not")
      . " a valid $tbl_name.$col_name value\n";

$dbh->disconnect ();

File: src/mysqlcookbook/recipes/transfer/check_set_value.py
#!/usr/bin/python3
# check_set_value.py: Wrapper to demonstrate check_set_value()
# utility function that determines whether a value is a legal
# value of a given SET column.

# Usage: check_set_value.py db_name tbl_name col_name test_value

import sys
import mysql.connector
import cookbook
from cookbook_utils import *

if len(sys.argv) != 5:
  print("Usage: check_enum_value.py db_name tbl_name col_name test_val")
  sys.exit(1)
db_name = sys.argv[1]
tbl_name = sys.argv[2]
col_name = sys.argv[3]
val = sys.argv[4]

try:
  conn = cookbook.connect()
except mysql.connector.Error as e:
  print("Error: %s" % e)
  sys.exit(1)

#@ _BULK_SET_CHECK_SETUP_
info = get_enumorset_info(conn, db_name, tbl_name, col_name)
members={}
# convert dictionary key to consistent lettercase
for v in info['values']:
  members[v.lower()] = 1
#@ _BULK_SET_CHECK_SETUP_

# If any element of the test value isn't in the SET,
# the entire value is invalid

#@ _BULK_SET_CHECK_TEST_
valid = 1 # assume valid until we find out otherwise
for v in val.split(","):
  if v.lower() not in members:
    valid = 0
    break
#@ _BULK_SET_CHECK_TEST_

print(val),
if valid:
  print("is"),
else:
  print("is not"),
print("a member of %s.%s" % (tbl_name, col_name))

conn.close()
